File: skilloper_app/lib/models/questionnaire.dart
class Question {
  final int id;
  final String questionType;
  final String question;
  final String? code;
  final String? language;
  final List<String> options;
  final int? correctAnswer; // For single_choice
  final List<int>? correctAnswers; // For multiple_choice
  final String? explanation;

  const Question({
    required this.id,
    required this.questionType,
    required this.question,
    this.code,
    this.language,
    required this.options,
    this.correctAnswer,
    this.correctAnswers,
    this.explanation,
  });

  factory Question.fromJson(Map<String, dynamic> json) {
    return Question(
      id: json['id'] as int,
      questionType: json['question_type'] as String? ?? 'single_choice',
      question: json['question'] as String,
      code: json['code'] as String?,
      language: json['language'] as String?,
      options: List<String>.from(json['options'] as List),
      correctAnswer: json['correctAnswer'] as int?,
      correctAnswers: json['correct_answers'] != null
          ? List<int>.from(json['correct_answers'] as List)
          : null,
      explanation: json['explanation'] as String?,
    );
  }


  bool get isMultipleChoice => questionType == 'multiple_choice';
  bool get isSingleChoice => questionType == 'single_choice';
}

class QuestionnaireSummary {
  final int id;
  final String title;
  final String description;
  final String type; // 'practice' or 'exam'
  final int maxOptions;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int questionCount;

  const QuestionnaireSummary({
    required this.id,
    required this.title,
    required this.description,
    required this.type,
    required this.maxOptions,
    required this.createdAt,
    required this.updatedAt,
    required this.questionCount,
  });

  factory QuestionnaireSummary.fromJson(Map<String, dynamic> json) {
    return QuestionnaireSummary(
      id: json['id'] as int,
      title: json['title'] as String,
      description: json['description'] as String,
      type: json['type'] as String,
      maxOptions: json['max_options'] as int,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      questionCount: json['question_count'] as int,
    );
  }

  bool get isPracticeMode => type == 'practice';

  int get estimatedMinutes => (questionCount * 1.5).ceil();
}

class Questionnaire {
  final int id;
  final String title;
  final String description;
  final String type; // 'practice' or 'exam'
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<Question> questions;

  const Questionnaire({
    required this.id,
    required this.title,
    required this.description,
    required this.type,
    required this.createdAt,
    required this.updatedAt,
    required this.questions,
  });

  factory Questionnaire.fromJson(Map<String, dynamic> json) {
    return Questionnaire(
      id: json['id'] as int,
      title: json['title'] as String,
      description: json['description'] as String,
      type: json['type'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      questions: (json['questions'] as List)
          .map((q) => Question.fromJson(q as Map<String, dynamic>))
          .toList(),
    );
  }


  bool get isPracticeMode => type == 'practice';
  
  int get estimatedMinutes => (questions.length * 1.5).ceil();
}

class ImportResponse {
  final String message;
  final QuestionnaireSummary questionnaire;

  const ImportResponse({
    required this.message,
    required this.questionnaire,
  });

  factory ImportResponse.fromJson(Map<String, dynamic> json) {
    return ImportResponse(
      message: json['message'] as String,
      questionnaire: QuestionnaireSummary.fromJson(json['questionnaire'] as Map<String, dynamic>),
    );
  }
}

class QuizResult {
  final int score;
  final int correct;
  final int total;
  final List<QuestionResult> questionResults;

  const QuizResult({
    required this.score,
    required this.correct,
    required this.total,
    required this.questionResults,
  });

  double get percentage => (correct / total) * 100;
}

class QuestionResult {
  final Question question;
  final int? userAnswer; // For single_choice
  final List<int>? userAnswers; // For multiple_choice
  final bool isCorrect;

  const QuestionResult({
    required this.question,
    this.userAnswer,
    this.userAnswers,
    required this.isCorrect,
  });
}

File: skilloper-api/internal/models/questionnaire.go
package models

import (
	"time"
)

// Database Models - Question types
const (
	QuestionTypeSingleChoice   = "single_choice"
	QuestionTypeMultipleChoice = "multiple_choice"
)

// Note: Limit constants (MinOptionsLimit, MaxOptionsLimit, etc.) are in limits.go

type Questionnaire struct {
	ID          uint       `json:"id" gorm:"primaryKey"`
	Title       string     `json:"title" gorm:"not null"`
	Description string     `json:"description"`
	Type        string     `json:"type" gorm:"not null;default:'practice'"` // "practice" or "exam"
	MaxOptions  int        `json:"max_options" gorm:"not null;default:4"`    // Maximum options per question
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
	Questions   []Question `json:"questions" gorm:"foreignKey:QuestionnaireID"`
}

type Question struct {
	ID                   uint      `json:"id" gorm:"primaryKey"`
	QuestionnaireID      uint      `json:"questionnaire_id"`
	QuestionType         string    `json:"question_type" gorm:"not null;default:'single_choice'"` // "single_choice", "multiple_choice"
	QuestionText         string    `json:"question" gorm:"not null"`
	AlternativeQuestions string    `json:"alternative_questions"` // JSON array of alternative question texts
	Code                 string    `json:"code"`
	Language             string    `json:"language"`
	Options              string    `json:"options"`             // JSON array stored as string
	AlternativeOptions   string    `json:"alternative_options"` // JSON array of additional options for variety
	CorrectAnswer        int       `json:"correctAnswer"`       // Index in the original options array (for single_choice)
	CorrectAnswers       string    `json:"correct_answers"`     // JSON array of indices (for multiple_choice)
	AlternativeAnswers   string    `json:"alternative_answers"` // JSON array of alternative correct answer texts
	Explanation          string    `json:"explanation"`         // Explanation for the correct answer
	CreatedAt            time.Time `json:"created_at"`
	UpdatedAt            time.Time `json:"updated_at"`
}

// Request DTOs
type CreateQuestionnaireRequest struct {
	Title       string            `json:"title"`
	Description string            `json:"description"`
	Type        string            `json:"type"` // "practice" or "exam"
	MaxOptions  int               `json:"max_options,omitempty"` // Optional, defaults to 4 if 0
	Questions   []QuestionRequest `json:"questions"`
}

type QuestionRequest struct {
	QuestionType         string   `json:"question_type,omitempty"` // "single_choice", "multiple_choice"
	Question             string   `json:"question"`
	AlternativeQuestions []string `json:"alternative_questions,omitempty"` // Alternative question texts
	Code                 string   `json:"code,omitempty"`
	Language             string   `json:"language,omitempty"`
	Options              []string `json:"options,omitempty"`
	AlternativeOptions   []string `json:"alternative_options,omitempty"` // Additional options for variety
	CorrectAnswer        int      `json:"correctAnswer,omitempty"`       // For single_choice (1-indexed: 1, 2, 3, 4)
	CorrectAnswers       []int    `json:"correct_answers,omitempty"`     // For multiple_choice (1-indexed: [1, 3, 4])
	AlternativeAnswers   []string `json:"alternative_answers,omitempty"` // Alternative correct answer texts
	Explanation          string   `json:"explanation,omitempty"`
}

// Response DTOs
type QuestionnaireSummary struct {
	ID             uint      `json:"id"`
	Title          string    `json:"title"`
	Description    string    `json:"description"`
	Type           string    `json:"type"`
	MaxOptions     int       `json:"max_options"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
	QuestionCount  int       `json:"question_count"`
}

type QuestionnaireResponse struct {
	ID          uint               `json:"id"`
	Title       string             `json:"title"`
	Description string             `json:"description"`
	Type        string             `json:"type"`
	MaxOptions  int                `json:"max_options"`
	CreatedAt   time.Time          `json:"created_at"`
	UpdatedAt   time.Time          `json:"updated_at"`
	Questions   []QuestionResponse `json:"questions"`
}

type QuestionResponse struct {
	ID             uint     `json:"id"`
	QuestionType   string   `json:"question_type"`
	Question       string   `json:"question"`
	Code           string   `json:"code,omitempty"`
	Language       string   `json:"language,omitempty"`
	Options        []string `json:"options,omitempty"`
	CorrectAnswer  int      `json:"-"` // Hidden from client - server validates answers
	CorrectAnswers []int    `json:"-"` // Hidden from client - server validates answers
	Explanation    string   `json:"explanation,omitempty"`
}

// Health check response
type HealthResponse struct {
	Status    string `json:"status"`
	Message   string `json:"message"`
	Timestamp string `json:"timestamp"`
}
